### PR TITLE
gh-117389: Fix `test_compileall.EncodingTest`

### PR DESCRIPTION
**Pull request by sobolevn**
Changes:
- Removed `coding`, because we don't really need that anymore
- Added `()` to `print`
- Now simulating the core issue instead: `bytes` cannot contain unicode chars
- Now checking the result of `compile_dir` call
- Now checking the stdout to contain what we want

@vstinner can you please take a look? :)

* Issue: gh-117389

### CODE REVIEW

File: Lib/test/test_compileall.py [hunk 1]
```diff
@@ -502,19 +502,19 @@ def setUp(self):
...
-            file.write('print u"\u20ac"\n')
+            file.write('print(b"\u20ac")\n')  # intentional error
 
     def tearDown(self):
         shutil.rmtree(self.directory)
 
     def test_error(self):
-        try:
-            orig_stdout = sys.stdout
-            sys.stdout = io.TextIOWrapper(io.BytesIO(),encoding='ascii')
```
Comment (vstinner) [inline Lib/test/test_compileall.py:514]: The bug report was that the code failed depending on stdout encoding: https://github.com/python/cpython/issues/50965#issuecomment-1093477751

> UnicodeEncodeError occurs depending on code page.

With your change, you don't control stdout encoding anymore.

File: Lib/test/test_compileall.py [hunk 2]
```diff
@@ -502,19 +502,19 @@ def setUp(self):
         self.directory = tempfile.mkdtemp()
         self.source_path = os.path.join(self.directory, '_test.py')
         with open(self.source_path, 'w', encoding='utf-8') as file:
-            file.write('# -*- coding: utf-8 -*-\n')
-            file.write('print u"\u20ac"\n')
+            file.write('print(b"\u20ac")\n')  # intentional error
```
Comment (vstinner) [inline Lib/test/test_compileall.py]: Please elaborate why it's an error, it's non obvious to me. Apparently, you expect the error:

```
SyntaxError: bytes can only contain ASCII literal characters
```

Example:

```suggestion
            # Intentional syntax error: bytes can only contain
            # ASCII literal characters.
            file.write('print(b"\u20ac")\n')
```

File: Lib/test/test_compileall.py [hunk 3]
```diff
@@ -502,19 +502,19 @@ def setUp(self):
...
-            compileall.compile_dir(self.directory)
-        finally:
-            sys.stdout = orig_stdout
+        buffer = io.StringIO()
+        with contextlib.redirect_stdout(buffer):
+            compiled = compileall.compile_dir(self.directory)
+        self.assertFalse(compiled)  # should not be successful
+        res = buffer.getvalue()
+        self.assertIn('SyntaxError', res)
+        self.assertIn('print(b"€")', res)
```
Comment (vstinner) [inline Lib/test/test_compileall.py]: I expect expect the string \u20ac, rather than the € character, if stdout encoding is ASCII.

**Comment by sobolevn** [comment]
I've addressed all your points, @vstinner! Agree with all of them. Please, take another look when you will have time.

File: Lib/test/test_compileall.py [hunk 4]
```diff
@@ -502,19 +502,25 @@ def setUp(self):
...
-        res = buffer.getvalue()
-        self.assertIn('SyntaxError', res)
-        self.assertIn('print(b"€")', res)
+        buffer.seek(0)
+        res = buffer.read()
+        self.assertIn(
+            'SyntaxError: bytes can only contain ASCII literal characters',
+            res,
+        )
+        self.assertNotIn('UnicodeEncodeError', res)
```
Comment (vstinner) [inline Lib/test/test_compileall.py:523]: I don't think that this check is useful.

**Review by vstinner — APPROVED** [review]
LGTM. You might remove UnicodeEncodeError test, or keep it. It's up to you.

File: Lib/test/test_compileall.py (hunk 4 above)
Comment (sobolevn) [reply Lib/test/test_compileall.py:523]: It was mentioned in https://github.com/python/cpython/issues/50965#issuecomment-1093477751 so, I guess that it won't hurt.

**Pull request merged**